crates/ra_hir/src/ids.rs

```diff
@@ -136,27 +139,48 @@ pub(crate) fn macro_def_query(db: &impl AstDatabase, id: MacroDefId) -> Option<A
 
 pub(crate) fn macro_arg_query(db: &impl AstDatabase, id: MacroCallId) -> Option<Arc<tt::Subtree>> {
     let loc = id.loc(db);
-    let macro_call = loc.ast_id.to_node(db);
-    let arg = macro_call.token_tree()?;
-    let (tt, _) = mbe::ast_to_token_tree(&arg)?;
-    Some(Arc::new(tt))
+
+    match loc {
+        MacroCallLoc::Macro { ast_id, .. } => {
+            let macro_call = ast_id.to_node(db);
+            let arg = macro_call.token_tree()?;
+            let (tt, _) = mbe::ast_to_token_tree(&arg)?;
+            Some(Arc::new(tt))
+        }
+
+        MacroCallLoc::Derive { .. } => None
+    }
 }
 
 pub(crate) fn macro_expand_query(
     db: &impl AstDatabase,
     id: MacroCallId,
 ) -> Result<Arc<tt::Subtree>, String> {
     let loc = id.loc(db);
-    let macro_arg = db.macro_arg(id).ok_or("Fail to args in to tt::TokenTree")?;
 
-    let macro_rules = db.macro_def(loc.def).ok_or("Fail to find macro definition")?;
-    let tt = macro_rules.expand(&macro_arg).map_err(|err| format!("{:?}", err))?;
-    // Set a hard limit for the expanded tt
-    let count = tt.count();
-    if count > 65536 {
-        return Err(format!("Total tokens count exceed limit : count = {}", count));
+    match loc {
+        MacroCallLoc::Macro { ast_id, def } => {
+            let macro_arg = db.macro_arg(id).ok_or("Fail to args in to tt::TokenTree")?;
+
+            let macro_rules = db.macro_def(def).ok_or("Fail to find macro definition")?;
+            let tt = macro_rules.expand(&macro_arg).map_err(|err| format!("{:?}", err))?;
+            // Set a hard limit for the expanded tt
+            let count = tt.count();
+            if count > 65536 {
+                return Err(format!("Total tokens count exceed limit : count = {}", count));
+            }
+            Ok(Arc::new(tt))
+        },
+
+        MacroCallLoc::Derive { ast_id} => {
+            // TODO: produce token trees of derive macro expansion
```

Comment:
@matklad can you please give a suggestion on how do we produce code (token trees) for expansion of `#[derive(...)]` attributes? Don't we have something similar to `quote!` macro for convenient token trees generation?

Reply:
Nope, we don't have `quote`, I am afraid one have to manually wrangle the token trees at the moment :(